[API] Change representation of custom fields

NobodysNightmare opened the pull request:
## OpenProject work packages
- https://community.openproject.org/work_packages/18296
- https://community.openproject.org/work_packages/18295
- https://community.openproject.org/work_packages/18294
- https://community.openproject.org/work_packages/18293
- https://community.openproject.org/work_packages/18292
- https://community.openproject.org/work_packages/18291
- https://community.openproject.org/work_packages/18290
- https://community.openproject.org/work_packages/18288
## Progress
- [x] schema for basic custom fields
- [x] schema for linked custom fields
- [x] schema for list custom fields
- [x] basic custom fields in WP representer
- [x] linked custom fields in WP representer
- [x] list custom fields in WP representer
- [x] patchable basic custom fields
- [x] patchable linked custom fields
- [x] patchable list custom fields
- [x] render text custom fields as formattables (forced to `plain`)
- [x] full specs for `CustomFieldInjector`
- [x] don't override `new`, but make a class creator that is usable by lambdas passed to `:decorator`
- [x] clean up messy bits and pieces >:D
- [x] validation of allowed values
## Progress during review
- [x] embed custom values of type user and version
- [x] find the root cause of unspecific custom value errors (they are more specific normally)
- [x] rename `wp_schema` of `inject_schema` (and probably adapt behaviour) to make method work for other customizables
- [x] specs for `ResourceLinkParser#parse_id`'s `property` parameter
- [x] check user deletion
- [x] avoid deep nesting in acts as custom...
- [x] fix test with journal stuff
- [x] falsey
- [x] method missing awesomeness
- [x] adapt **frontend** to missing `customProperties` dictionary on `/api/v3/work_packages/:id`
- [x] adapt StringObject

NobodysNightmare committed:
add schema information for basic custom fields

NobodysNightmare committed:
override :new to create schema representers

this is quite evil... I chose this solution,
because it encapsulates the problem quite well:
- no knowledge about any factory method required
- compatible with automatic instantiation by ROAR
- "it just works" (until it doesn't)

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://feature-api-custom-fields-in-schema.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://feature-api-custom-fields-in-schema.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare committed:
new custom value representation in WP representer

- also allows writing custom values

NobodysNightmare committed:
reduce duplication inside WP schema representer

NobodysNightmare committed:
allow injector to inject allowed values

NobodysNightmare committed:
use shared examples for tests

NobodysNightmare committed:
specs for all paths of :inject_schema

NobodysNightmare committed:
fix value injection for nil values

NobodysNightmare committed:
style

NobodysNightmare committed:
slightly improve readability

NobodysNightmare committed:
enable linked properties for WP

- disabling wrong representation in payload

NobodysNightmare committed:
enable patching of linked properties

NobodysNightmare committed:
fix patching of empty string objects

NobodysNightmare committed:
use factory method for _links representer

this will never need to be instantiated by ROAR

NobodysNightmare committed:
rewrite ResourceLinkParser

- use own RegExp instead of borrowing the ones from Grape routes
- add specs

NobodysNightmare committed:
extract method into ResourceLinkParser

- similar code was already used in two different places
- now the error message makes sense for both occurences

NobodysNightmare committed:
fix specs

NobodysNightmare committed:
remove obsolete specs

- were disabled (and outdated) anyway
- are both already implemented elsewhere in the same test file

NobodysNightmare committed:
validate user and version custom fields

NobodysNightmare committed:
style

NobodysNightmare committed:
add another case to repair list custom fields

Genius idea to require the parameter of a method to carry totally
different information, depending on internal state of the object

NobodysNightmare committed:
update method doc

NobodysNightmare committed:
stop overriding new

NobodysNightmare commented on lib/api/v3/work_packages/work_packages_api.rb:
```diff
@@ -78,8 +78,8 @@ def write_request_valid?
-              @representer = WorkPackages::WorkPackageRepresenter.new(work_package,
-                                                                      current_user: current_user)
+              @representer = WorkPackages::WorkPackageRepresenter.create(work_package,
+                                                                         current_user: current_user)
```
100 is close enough, pet!

-.-

NobodysNightmare committed:
make a formattable of text custom fields

NobodysNightmare committed:
add more specs for the CF Injector

- fixing the worst typo ever
- failing for bool, float and int...

NobodysNightmare committed:
let custom values return typed values

this allows to have type-correct JSON output, e.g.:

12 instead of "12"
true instead of "1"

NobodysNightmare committed:
fix broken specs

NobodysNightmare committed:
add remaining specs for CF injector

NobodysNightmare committed:
sty13

NobodysNightmare committed:
add required and writable into parameter

- are set on all instance creations anyway
- still optional in case you do not want to set it

ulferts commented on app/models/custom_value.rb:
```diff
@@ -38,6 +38,32 @@ class CustomValue < ActiveRecord::Base
+    when 'user'
+      User.find(value)
+    when 'version'
+      Version.find(value)
```
This will cause 500 errors in cases where a version/user is deleted after a value has been set. For users, we should double check whether user deletion does rewrite such values to the DeletedUser. For versions, no such mechanism exists. We might be forced to add it now.

NobodysNightmare replied:
For user I'd think the deleted user is fine and we should double check.

For the version I'd say `find_by_id` to the rescue!

ulferts commented on app/models/custom_value.rb:
```diff
@@ -38,6 +38,32 @@ class CustomValue < ActiveRecord::Base
+    return nil if value.nil?
+
+    # strings may be blank, all other types can only be nil if their value is blank
+    return nil if value.blank? && !(['string', 'text'].include?(custom_field.field_format))
```
why not make that part of the chase statement below?

ulferts replied:
forget it, wrote before thinking.

NobodysNightmare committed:
extract some case statements into fancy strategies

- 17 new files
- lots of new lines of code
- specs!
- same functionality as before

NobodysNightmare commented on app/models/custom_value/string_strategy.rb:
```diff
@@ -0,0 +1,37 @@
+
+class CustomValue::StringStrategy < CustomValue::FormatStrategy
+  def typed_value
+    return value
```
But it looks so much nicer!

0xF013 replied:
@NobodysNightmare, the dog whisperer

NobodysNightmare committed:
only perform validation if it is possible

IMO the best way to deal with that ugly inter-validation dependency.
You just can't meaningfully test whether a value is valid,
when you don't know its format.

NobodysNightmare commented:
Tests look green to me

ulferts commented on app/models/custom_field.rb:
```diff
@@ -101,6 +101,10 @@ def validate_presence_of_possible_values
   # validate default value in every translation available
   def validate_default_value_in_translations
+    # it is not possible to determine the validity of a value, when there is no valid format.
+    # another validation will take take of adding an error, but here we need to abort
```
> another validation will take take of adding an error, but here we need to abort

spelling

ulferts commented on app/models/custom_value.rb:
```diff
@@ -79,16 +97,9 @@ def validate_format_of_value
-        errors.add(:value, :inclusion) unless custom_field.possible_values.include?(value)
+      result = strategy.validate_type_of_value
+      if result
+        errors.add(:value, result)
```
This is ether a question of naming or of correctness. 

If it is just a question of naming you should rename this variable 'error' instead of 'result' because you are expecting a single error to be returned or nothing at all. 

If it is a question of correctness, then I'd say the implementation is wrong :). You do not know how many errors might be returned. You could either move the whole error setting to within the strategy. I think that is ok because the strategy is a part of the custom value.  Or you accept multiple error to be returned. IMO, assigning the errors within the strategy is preferable.

NobodysNightmare replied:
I liked keeping the errors interface out of the strategy... Indeed I thought about returning a list, but it was never neccessary I so I thought I'd keep it simple and if someone needs it, he will build it.

ulferts commented on app/models/custom_value.rb:
```diff
@@ -99,4 +110,10 @@ def validate_length_of_value
+  private
+
+  def strategy
+    @strategy = FORMAT_STRATEGIES[custom_field.field_format].new(self)
```
Every call to strategy will instantiate a new strategy. What you mean to do is:
`@strategy ||= FORMAT_STRATEGIES[custom_field.field_format].new(self)`

NobodysNightmare replied:
Actually that is really what I meant to do... as in: I wanted to write `||=` and when reading your comment I could at first not see what you thought was wrong ^^

ulferts commented on app/models/custom_value/bool_strategy.rb:
```diff
@@ -0,0 +1,39 @@
+
+class CustomValue::BoolStrategy < CustomValue::FormatStrategy
+  def typed_value
+    unless value.nil? || value.blank?
```
`unless value.blank?`

does suffice

ulferts commented on app/models/custom_value.rb:
```diff
@@ -44,6 +56,12 @@ def set_default_value
+
   # Returns true if the boolean custom value is true
   def true?
     self.value == '1'
```
This should be handled by the BoolStrategy now.

NobodysNightmare replied:
As in:

``` ruby
self.typed_value == true
```

or an explicitly implemented method for all strategies?

I should probably see where this method is used and just remove it...

ulferts commented on app/models/custom_value/bool_strategy.rb:
```diff
@@ -0,0 +1,39 @@
+  end
+
+  def validate_type_of_value
+  end
```
So it is correct to set the value to "Pustekuchen"?

NobodysNightmare replied:
Indeed!

`Pustekuchen` corresponds to `false`... Same for `Unicorn`.

I was not sure how safe it is to introduce such a validation, when there was no validation in the old code (breaking existing stuff).

But strictly speaking you are right and I should probably also restrict it.

ulferts commented on app/models/custom_value/date_strategy.rb:
```diff
@@ -0,0 +1,43 @@
+
+class CustomValue::DateStrategy < CustomValue::FormatStrategy
+  def typed_value
+    unless value.nil? || value.blank?
```
please see above

ulferts commented on app/models/custom_value/float_strategy.rb:
```diff
@@ -0,0 +1,43 @@
+
+class CustomValue::FloatStrategy < CustomValue::FormatStrategy
+  def typed_value
+    unless value.nil? || value.blank?
```
please see above

ulferts commented on app/models/custom_value/date_strategy.rb:
```diff
@@ -0,0 +1,43 @@
+    Date.iso8601(value)
+    return nil
+  rescue
+    return :not_a_date
```
You don't really need explicit return statements here. This is no JavaScript.

NobodysNightmare replied:
Is it more pretty to not use return statements? I fear it is...

will change it everywhere...

ulferts commented on app/models/custom_value/format_strategy.rb:
```diff
@@ -0,0 +1,56 @@
+  # Returns the value of the CustomValue in a typed fashion (i.e. not as the string
+  # that is used for representation in the database)
+  def typed_value
+    raise 'SubclassResponsibility'
```
This is typically done with a [NotImplementedError](http://ruby-doc.org//core-2.2.0/NotImplementedError.html)

NobodysNightmare replied:
I don't agree:
1. I asked the internet and did not find an example using this exception
2. The documentation that you link clearly states a different intention:

> Raised when a feature is not implemented **on the current platform**. For example, methods depending on the fsync or fork system calls may raise this exception if the **underlying operating system or Ruby runtime** does not support them.

(highlighting by me)

ulferts commented on app/models/custom_value/format_strategy.rb:
```diff
@@ -0,0 +1,56 @@
+
+  def value
+    custom_value.value
+  end
```
It is shorter if you use [delegate](http://api.rubyonrails.org/classes/Module.html#method-i-delegate)

ulferts commented on app/models/custom_value/int_strategy.rb:
```diff
@@ -0,0 +1,43 @@
+
+class CustomValue::IntStrategy < CustomValue::FormatStrategy
+  def typed_value
+    unless value.nil? || value.blank?
```
please see above

ulferts commented on app/models/custom_value/list_strategy.rb:
```diff
@@ -0,0 +1,42 @@
+
+class CustomValue::ListStrategy < CustomValue::FormatStrategy
+  def typed_value
+    unless value.nil? || value.blank?
```
By now you know what I would write here.

ulferts commented on app/models/custom_value/user_strategy.rb:
```diff
@@ -0,0 +1,42 @@
+
+class CustomValue::UserStrategy < CustomValue::FormatStrategy
+  def typed_value
+    unless value.nil? || value.blank?
```
Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.

ulferts commented on app/models/custom_value/user_strategy.rb:
```diff
@@ -0,0 +1,42 @@
+class CustomValue::UserStrategy < CustomValue::FormatStrategy
+  def typed_value
+    unless value.nil? || value.blank?
+      User.find_by_id(value)
```
@NobodysNightmare can you please check whether the user deletion rewires custom fields of format user correctly?

ulferts commented on app/models/custom_value/version_strategy.rb:
```diff
@@ -0,0 +1,42 @@
+
+class CustomValue::VersionStrategy < CustomValue::FormatStrategy
+  def typed_value
+    unless value.nil? || value.blank?
```
![image](https://cloud.githubusercontent.com/assets/617519/6393083/604f911c-bdc7-11e4-9711-7220fada4483.png)

NobodysNightmare committed:
various small fixes

ulferts commented on lib/api/decorators/schema.rb:
```diff
@@ -0,0 +1,128 @@
+                        required: true,
+                        writable: true,
+                        min_length: nil,
+                        max_length: nil)
```
Those indentations are weird IMO. Hound, where are you when I need you!

NobodysNightmare replied:
My IDE is not very supportive here... Happened when I removed the `self.` prefix.

NobodysNightmare committed:
remove unneccessary return statements

this is how ruby style works... apparently

ulferts commented on lib/api/decorators/schema.rb:
```diff
@@ -0,0 +1,128 @@
+
+                     representer
+                   }
+        end
```
The only difference between the `schema_with_allowed_link` and `schema_with_allowed_collection` method is in the getter block. So you could write a private method takes the block as a parameter. This should really be easier to read.

NobodysNightmare replied:
I don't quite see your plan.

The major part of both methods **is** actually the getter block, so what you are saying is basically "the methods are equal except most of their content".

I also don't see how I would do it, because both methods also differ in their signature.

So when extracting a private method I would basically end up with a the same methods as before, but I don't need to set `exec_context: :decorator`.

ulferts replied:
The only difference is the representer used so the difference is:

```
::API::Decorators::AllowedValuesByLinkRepresenter.new(
                       type: type,
                       name: title,
                       required: required,
                       writable: writable)
```

vs.

```
::API::Decorators::AllowedValuesByCollectionRepresenter.new(
                       type: type,
                       name: title,
                       current_user: current_user,
                       value_representer: value_representer,
                       link_factory: -> (value) { instance_exec(value, &link_factory) },
                       required: required,
                       writable: writable)
```

Therefore that is what you have to provide to a method that looks like this:

```

        def some_intelligent_name(property, representer)
          raise ArgumentError unless property

          property property,
                   exec_context: :decorator,
                   getter: -> (*) {
                     if represented.defines_assignable_values?
                       representer.allowed_values = instance_exec(&values_callback)
                     end

                     representer
                   }
        end
```

NobodysNightmare replied:
I'll have another look at it, but I still doubt it is as easy...

## Problems

**Assignment of allowed values different for both**

You say the shared part contains the assignment of the allowed values:

``` ruby
if represented.defines_assignable_values?
  representer.allowed_values = instance_exec(&values_callback)
end
```

However, strictly speaking the code is not identical. On one case a method `allowed_values` is called, in the other case the method is called `allowed_values_href`. That is because the datatype of both calls is significantly different (list of things vs. string).

_Possible workaround:_ Calling both methods the same, and expecting the caller to provide the correct result in the callback.

**Need for instance evaluation in representer**

The representers of `schema_with_allowed_collection` are created using `instance_exec` for their `link_factory`. I think the only reason for that is that we want to use `api_v3_paths`, which is only available in the context of the representer, but not in the context of its class.

_Possible workaround:_ Deviate from the usual method of including the `PathHelper`.
Option 1: I could reference `ApiV3Path` directly instead of `api_v3_paths`, probably inside a class method of my own.
Option 2: It might be possible to include the `PathHelper` module within my `class << self`. My Ruby foo is not good enough to know if this will work, but I would try it. However I still don't know if this would be a good or a bad idea.

**More need for representer context**

After another look at the representer created in `schema_with_allowed_collection`, I noticed that there is another need for code evaluation inside the representer context (as opposed to execution on the class level): We access the instance method called `current_user`, which is important to correctly create the embedded instances of the allowed values collection.

I don't know if it is strictly neccessary to have the `current_user` as instance variable of the representers. However, this is a design decision that was made before I arrived here and I think it would at least take some effort to remove the `current_user` instance method from all representers (which is what we needed to do consequently).

_Possible workaround:_ See above, I don't know if this was feasible.

_Generic workaround for both problems above:_ Add another callback that creates the representer for allowed values and also evaluate that one in the right context. But I doubt, that readability is something that can be improved by adding more layers of indirection.

**tl;dr:**

I think our main problem is that we need the context of the schema representer to create the allowed value representers.

NobodysNightmare replied:
An additional note to my short comment above:

> I don't know if it is strictly neccessary to have the `current_user` as instance variable of the representers. However, this is a design decision that was made before I arrived here and I think it would at least take some effort to remove the `current_user` instance method from all representers (which is what we needed to do consequently).

After seeing how stuff works, I am now in favor of removing the `current_user` from the representers. It seems as if it is already causing trouble:

``` ruby
# work_package_representer.rb
property :version,
         embedded: true,
         exec_context: :decorator,
         if: ->(*) { represented.fixed_version.present? }

def version
  Versions::VersionRepresenter.new(represented.fixed_version, current_user: current_user)
end
```

Because ROAR is not able to transfer the `current_user` automatically into the contained representer, this has to be done manually.

This is something that now causes headaches for me, too, as I am trying to embed the version of a custom field in a work package.

NobodysNightmare committed:
use fancy ruby shortcuts

NobodysNightmare commented:
@ulferts wrote:

> @NobodysNightmare can you please check whether the user deletion rewires custom fields of format user correctly?

This diff got outdated, but was not yet fixed

ulferts commented on app/models/custom_value/list_strategy.rb:
```diff
@@ -0,0 +1,42 @@
+  def typed_value
+    unless value.blank?
+      value
+    end
```
Thought experiment: What is the expected value of this if the value was set and then the list of allowed values is changed to no longer include the value that is already taken by the CustomValue.

NobodysNightmare replied:
Thought experiment No. 2:
You've got a string custom field that takes any value.
After some time you change it to only accept strings matching a specific regular expression.

What is your expected output for non-matching strings?

NobodysNightmare replied:
**Assumption 1:** Both thought experiments should come to the same conclusion
**Assumption 2:** For Thought experiment No. 2 I'd expect the wrong value being returned

**General conclusion:** I'd expect to see the "wrong" value

Additional argument: If you've got two locales for a list, you will also be able to receive a value that is not allowed in your locale.

NobodysNightmare committed:
style

must have gone wild when I removed the "self."

ulferts commented on lib/api/utilities/resource_link_parser.rb:
```diff
@@ -30,20 +30,65 @@
+        private
+
+        def resource_matcher
+          @@matcher ||= Regexp.compile(RESOURCE_REGEX)
```
You don't need a class variable here. A [class level instance variable](http://www.railstips.org/blog/archives/2006/11/18/class-and-instance-variables-in-ruby/) is enough and less complicated.

ulferts commented on lib/api/utilities/resource_link_parser.rb:
```diff
@@ -30,20 +30,65 @@
+      # N.B. valid characters for URL path segments as of
+      # http://tools.ietf.org/html/rfc3986#section-3.3
+      SEGMENT_CHARACTER = '(\w|[-~!$&\'\(\)*+\.,:;=@]|%[0-9A-Fa-f]{2})'
+      RESOURCE_REGEX = "/api/v(?<version>\\d)/(?<namespace>\\w+)/(?<id>#{SEGMENT_CHARACTER}*)\\z"
```
This will fail if ever we try to match nested resources, e.g.:

```
/api/v3/projects/5/members/7
```

Not sure if that is a likely scenario, though.

NobodysNightmare replied:
I know. However, my thought was that we said that we probably don't like nested resources like in the old days:

```
/api/v3/projects/5/work_packages/7
```

Your member example would be correct, if we would not link to a user.

However, to get my point home: The old implementation relied on a path segment called `":id"` being defined, which would also have failed (or not worked reliably) for those kinds of resources. So I am at least as good as that one ;-)

My hope is that we will never need nested links for resources we will allow setting via link.
That would also open a wide variety of new problems, e.g. setting `project/2/member/4` as member for project **3**.

**tl; dr:** Not a problem of today, not a new problem, _hopefully_ not a problem of tommorrow

NobodysNightmare committed:
better naming

ulferts commented on lib/api/utilities/resource_link_parser.rb:
```diff
@@ -30,20 +30,65 @@
+          version = "v#{version}" || ':apiVersion'
+          namespace = namespace || ':resource'
+          "/api/#{version}/#{namespace}/:id"
+        end
```
I have to admit, that I liked the replaced version better, because now we have to maintain our own link parsing, whereas we before relied on grape to handle it for us. The new code does not care whether the route truly exists, the old code did.

NobodysNightmare replied:
The new code does care, because we cared **anyway**.

The calling code (now mostly extracted into `parse_id`) checked that it was the _requested_ resource anyway. Because `/api/v3/versions/12` is a valid route, does not mean that I want to set my status to it.

The argument of grape checking that the route really exists only had any value, if we would not check that. But we do. Always.

However, the problem with letting grape do things is, that we rely on grape doing it the way that we expect.
And we really used **implementation details** of grape here (including `instance_variable_get`). And those implementation details are not on our side w.r.t. links like `/api/v3/string_objects/`. The regex would not match with the trailing slash present, though the route is valid.

I don't even want to get started on updating or removing grape (the latter one actually [being spiked](https://github.com/opf/openproject/pull/2352) ^^)

NobodysNightmare committed:
use class side instance variable

NobodysNightmare committed:
remove 'true?' helper from custom_value

ulferts commented on lib/api/v3/utilities/custom_field_injector.rb:
```diff
@@ -0,0 +1,240 @@
+
+        # N.B. accepting a wp_schema here is not too great, but seems like the best way
+        # to obtain available versions and users for WP custom fields
+        def inject_schema(custom_field, wp_schema: nil)
```
This will not only be limited to work package, right? Because User, Group, Version, ... also have custom fields.

ulferts commented on lib/api/v3/work_packages/form/work_package_attribute_links_representer.rb:
```diff
@@ -39,6 +39,23 @@ class WorkPackageAttributeLinksRepresenter < Roar::Decorator
+                injector.inject_patchable_link_value(custom_field)
+              end
+
+              klass.new(work_package)
```
This could probably be moved to the injector, because Version et al will also do something like this.

ulferts commented on lib/api/v3/work_packages/form/work_package_payload_representer.rb:
```diff
@@ -38,6 +38,27 @@ class WorkPackagePayloadRepresenter < Roar::Decorator
+                injector.inject_value(custom_field)
+              end
+
+              klass
```
This could probably be moved to the injector, because Version et al will also do something like this.

ulferts commented on lib/api/v3/work_packages/work_package_representer.rb:
```diff
@@ -34,6 +34,22 @@ module API
+            end
+
+            klass
+          end
```
And yet again something to be moved to the Injector. Please compare to `WorkPackageSchemaRepresenter`.

ulferts commented on spec/lib/api/utilities/resource_link_parser_spec.rb:
```diff
@@ -0,0 +1,146 @@
+        subject.parse_id('/api/v3/types/14', property: 'foo', expected_namespace: 'statuses')
+      }.to raise_error(::API::Errors::Form::InvalidResourceLink)
+    end
+  end
```
The specs do not explain what the `property:` parameter is used for.

ulferts commented on lib/api/v3/work_packages/schema/work_package_schema_representer.rb:
```diff
@@ -159,97 +133,59 @@ def self.schema_with_allowed_link(property,
+                                             href: api_v3_paths.priority(priority.id),
+                                             title: priority.name
+                                           }
+                                         }
```
This does look nicer :+1: Crazy... but nicer.

ulferts commented on lib/api/v3/work_packages/work_package_representer.rb:
```diff
@@ -338,13 +352,6 @@ def version
-          values.map do |v|
-            { name: v.custom_field.name, format: v.custom_field.field_format, value: v.value }
-          end
-        end
```
This will probably break the front end. @NobodysNightmare, shall we tell @0xF013 or shall we surprise him?

NobodysNightmare replied:
I think he likes surprises...

NobodysNightmare replied:
And if it doesn't work in production we'll just blame him :sunglasses:

0xF013 replied:
Whatever, I have two people in my office to blame. What was that 7xx error?

ulferts commented on spec/lib/api/v3/utilities/custom_field_injector_spec.rb:
```diff
@@ -0,0 +1,326 @@
+  end
+
+  describe '#inject_schema' do
+    let(:modified_class) { Class.new(::API::Decorators::Schema) }
```
The idea of splitting up the injector into one for schema and one for actual objects does get some support when I see this, because you are actually testing both classes here. The test expects json paths to match which is a functionality provided by ::API::Decorators::Schema.

Would it make sense in your opinion to abandon the injector (the schema part) and make it a part of the Schema because it actually relies on Schema's interface rather heavily.

ulferts commented:
As far as I have seen, there are no tests that guarantee, that the custom values are made part of the work package representation. Is that correct @NobodysNightmare?

NobodysNightmare committed:
move creation logic into injector class

helpers should be quite reusable across customizables

ulferts commented:
Two remarks regarding functionality:
1) Custom values for fields of type "User" and "Version" are not embedded. That is inconsistent to the rest of the linked properties.
2) If a custom field is erroneous, the error message returned is: "Custom fields is invalid". This is not very helpful as it will not allow a client to know which field caused the error especially if he modifies more than one CustomValue.

NobodysNightmare commented:
@ulferts I added empty checkboxes for things I did not already fix and I agree with you that I want to fix them ;-)

0xF013 commented:
ping me when I should update the frontend

NobodysNightmare committed:
extract current_user into Single

a current_user known by representers is an unfortunate concept,
that we are using across multiple representers, so we should include
it in the top level class for our representers.

NobodysNightmare committed:
embed CF versions, users and string objects

NobodysNightmare committed:
split method

NobodysNightmare committed:
use variable names

NobodysNightmare committed:
add custom value errors to customizable

- allows to relate error messages to their custom field
- allows to use errors.full_message correctly for custom values
- might break specs... we will see that soon

Azure7111 commented on lib/api/decorators/property_schema_representer.rb:
```diff
@@ -33,11 +33,11 @@
-        @required = true
-        @writable = true
+        @required = required
+        @writable = writable
```
You might want to consider adding `@regularExpression` for custom field validation.

NobodysNightmare replied:
you mean `@regular_expression` ;-)

NobodysNightmare committed:
forbid validation of associated custom values

this SHOULD already be the default, however specs failed here
because of additional validation that took place. Those validations
stopped by adding the validate-parameter. Better don't ask why I guess...

Development and Production mode behaved as documented though (validate false by default)

NobodysNightmare committed:
fix weird specs

probably only worked because of unintended validation error from automatic validation:

- testing to pass nil as value, though nil-values were explicitly ignored
- assuming a single journal entry, though the WP is saved twice

really needed to be fixed, because of new implementation:

- valdiation errors now indicate CF id and are more specific

NobodysNightmare committed:
fix expectations

NobodysNightmare committed:
fix failing test unit tests

requires to run validations for custom values every time, so we do it

there was no clear indication on why they should not run always, it was not apparent either

NobodysNightmare committed:
include more schema details for custom fields

NobodysNightmare committed:
avoid variable shadowing

though unproblematic in this case

NobodysNightmare commented on config/initializers/10-patches.rb:
```diff
@@ -73,39 +73,22 @@ def self.human_attribute_name(attr, options = {})
+      attr_name = @base.class.human_attribute_name(attribute, :default => attr_name)
+      I18n.t(:"errors.format", {
+                               :default   => "%{attribute} %{message}",
+                               :attribute => attr_name_override || attr_name,
```
This is reckless copy&paste code from `ActiveModel`. I will try to keep the diff here as small as possible.

NobodysNightmare committed:
add patch hint to full_message

this hopefully makes clear which parts were changed by us and which belong to ActiveModel

NobodysNightmare committed:
fix another broken spec

NobodysNightmare committed:
ignore custom values that are to be deleted

- fixing the last failing test

NobodysNightmare committed:
add methods to directly access custom fields

NobodysNightmare committed:
use direct CF accessors in CF injector

NobodysNightmare committed:
some renaming

we are generifying the interpretation of the parameter we receive...

added some words about assumptions...

NobodysNightmare committed:
test inclusion of property name

ulferts commented on lib/api/errors/validation.rb:
```diff
@@ -33,21 +33,21 @@ class Validation < ErrorBase
           end
 
-          hash[key] = ::API::Errors::Validation.new(messages)
+          hash[attribute.to_s.camelize(:lower)] = ::API::Errors::Validation.new(messages)
```
Didn't know camelize has parameters - learned something today. Thanks @NobodysNightmare.

ulferts commented on config/initializers/10-patches.rb:
```diff
@@ -73,39 +73,25 @@ def self.human_attribute_name(attr, options = {})
+                               :attribute => attr_name_override || attr_name,
+                               :message   => message
+                             })
     end
```
I am aware that you are copying and then extending the method. I believe this is a good use case for [Module#prepend](http://gshutler.com/2013/04/ruby-2-module-prepend/) because you then only have to overwrite the overwriting of the attribute name. 

That would enable you to do:

<pre>

module CustomFieldErrorMessages
    def full_message(attribute, message)
      match = /\Acustom_field_(?<id>\d+)\z/.match(attribute)

      if match
        attribute = CustomField.find_by_id(match[:id]).name
      end

      super
    end
end

ActiveRecord::Errors.prepend CustomFieldErrorMessages

</pre>


At least I think this should be possible. Never tried it.

ulferts replied:
The module could even be moved within ActsAsCustomizable to reduce the gems' spread.

NobodysNightmare replied:
I agree with you that we should move that away from `10-patches` (I bet there are more than only ten patches included ;-) ).

However:

```
attr_name = attribute.to_s.gsub('.', '_').humanize
```

will probably dehumanize our well formatted attribute names...

ulferts commented on spec/models/work_package/work_package_custom_fields_spec.rb:
```diff
@@ -75,45 +75,49 @@ def self.change_custom_field_value(work_package, value)
               subject { work_package.save }
 
-              it { is_expected.to be_falsey }
+              it { is_expected.to be_false }
```
@NobodysNightmare why the renaming? be_false is deprecated AFAIK.

NobodysNightmare replied:
Indeed it is... What I **wanted** to do was `be false` (mind the space), however... After asking the internet I am all more ruby when I just ask for falsiness...

ulferts commented on spec/models/work_package/work_package_custom_fields_spec.rb:
```diff
@@ -256,7 +260,7 @@ def self.change_custom_field_value(work_package, value)
         end
 
-        subject { work_package.journals.first.customizable_journals.first.value }
+        subject { work_package.journals.last.customizable_journals.first.value }
```
There is no default order defined for journals. So the order in which the journals are returned is determined in the DB. You might be lucky of course but then again, you might not be.

ulferts commented on lib/plugins/acts_as_customizable/lib/acts_as_customizable.rb:
```diff
@@ -116,6 +119,35 @@ def reset_custom_values!
+              self.custom_field_values = { custom_field.id => value }
+            end
+          end
+        end
```
I have thought about this again and I fear that we might add a performance impediment here. Imagine you get a list of WorkPackages. For every WorkPackage the list of CustomFields would have to be fetched from the DB. That cannot be fast.

So I'd now opt for the method_missing variant you favoured initially already. We can still add the method once it has been found missing so the code is not lost.

NobodysNightmare commented:
@0xF013 Feel free to adapt the frontend, if there are any questions don't ~~bother~~ hesitate asking.

I've prepared a checkbox in the description for you. Feel free to check it whenever you like

edit: please **bother** asking! don't **hesitate**! **_WORDS**_!

0xF013 commented:
@NobodysNightmare 

> don't bother asking

NobodysNightmare committed:
only create CF accessor methods on demand

NobodysNightmare committed:
change expectation back to falsey

be_false == be_falsey, both expect falsey results
alternative would be "be false", which expects real false...

*confused*

NobodysNightmare committed:
simplify loop

NobodysNightmare committed:
don't pick the (random) last journal

pick the first (and hopefully only) journal that has customizable
journals

0xF013 committed:
switch frontend to new format

0xF013 committed:
hound

0xF013 committed:
fix some formatting

0xF013 committed:
fix double quotes

0xF013 committed:
fix protractor mocks

NobodysNightmare commented:
~~I possibly found an error related to this branch and need to investigate:
Seems that I am unable to change the type of a work package if that would remove a custom field of type user.~~

I am not able to reproduce it, so it never happened...

0xF013 committed:
hack-ish fix for stringobject

NobodysNightmare commented:
Trying to review on the JS-side of life...

NobodysNightmare commented on frontend/app/work_packages/controllers/details-tab-overview-controller.js:
```diff
@@ -137,9 +137,42 @@ module.exports = function($scope,
+    }
+
+    function getCustomProperties(workPackage) {
+      return _.compact(_.map(
```
@0xF013: Do @houndci and me just not understand your _m4d J4v4Scr1p7 Sk1llz_?

Or is this really an error?

0xF013 replied:
@NobodysNightmare '_' is the god object of a lib called lodash that does all that fancy ruby stuff you have like .flatten, .compact, .map, .reduce. This var is global and hound does not see it. I am sure there is some weird /\* iddqd */ comment that would make it shut up

NobodysNightmare commented on frontend/app/work_packages/controllers/details-tab-overview-controller.js:
```diff
@@ -137,9 +137,49 @@ module.exports = function($scope,
+
+    function getValue(workPackage, prop, propName) {
+      if (workPackage.props[propName]) {
+        return workPackage.props[propName].raw;
```
For most custom field types you want to directly access `workPackage.props[propName]`, e.g. `String`, `Integer`, `Float`, etc.

It is only `Formattable` that has a `raw` attribute. Like with description the meaning of the subproperties is:
- `format`: The format used to transform from raw to html (not that interesting for most clients I guess)
- `raw`: That is the string to display in an **editing** interface and send when editing... It will be transformed into html according to the format
- `html`: An already prepared html string, should be ready to use out of the box

**tl; dr:**
This line:
1. only works for custom fields of type `Formattable` ("Long Text" in the web interface), not for `Integer` et al
2. you should probably use the `html` subelement instead of `raw`, because otherwise line breaks are not rendered correctly

NobodysNightmare commented on frontend/app/work_packages/controllers/details-tab-overview-controller.js:
```diff
@@ -137,9 +137,49 @@ module.exports = function($scope,
+    function getFormat(workPackage, prop, propName) {
+      return workPackage.props[propName] ?
+              workPackage.props[propName].format :
+              prop.type.toLowerCase();
```
I am not sure where you are ultimately using the result of this method (for me it ends at line 181, where each "customProperty" is `push`ed into an abyss), however I believe you always want to use `prop.type`. 

`workPackage.props[propName].format` would either give you `plain` or `textile` (me knowing the implementation for custom fields: as of today always `plain`). Opposed to that `prop.type` will give you things like `String`, `Formattable`, `Integer`, `Version`, etc.

0xF013 committed:
fix primitive CF types

0xF013 commented:
@NobodysNightmare I stayed with the raw instead of html for now since I'd have to do some magic around rendering it. Is it important that it's rendered as HTML now?

NobodysNightmare commented:
Not too important... as it is no textile we are losing nothing like **bold** or _italic_.

However, we will be losing line breaks for now. So you should at least keep that in mind (or in a ticket).

edit: or in a TODO-comment, I don't know

0xF013 commented:
@NobodysNightmare it will be fixed automatically when the fields will become editable

NobodysNightmare commented:
@0xF013: The karma tests failed three times in a row. This does not seem to be good, probably they are really failing.

This could be related to the fact that I am unable to open the details pane on teatro.

0xF013 commented:
@NobodysNightmare bummer, it's green locally

0xF013 committed:
fix karma fail

0xF013 committed:
Merge branch 'feature/api_custom_fields_in_schema' of https://github.com/opf/openproject into feature/api_custom_fields_in_schema

0xF013 committed:
actually xit the test since it will become obsolete once the details pane is rewritten when making all inplace editable
